Decode values in most Elements; improve efficiency; update README.md

OllieJones opened the pull request:
As part of a project where I need to decode webm files (generated by the [MediaDevices](https://developer.mozilla.org/en-US/docs/Web/API/MediaDevices)  interface in Google Chrome and Firefox), I added some capabilities to node-ebml.

- It now interprets the payload in various Element data types, and includes a `value` member in the `tagObj`.
- It unpacks `Block` and `SimpleBlock` elements (see the README for details).
- It uses `TypedArray` data throughout, and uses `.subarray()` instead of `.slice()` operations to reduce the copying of data and so improve performance.

themasch commented:
Hi @OllieJones, thanks for the massiv addition to the project. 
Could you please have a look at the tests & lints, since the builds are currently failing. 

The first reason for this is eslint failing. I've looked at the list of lints that failed and I guess we have to tune some rules. Additionally we currently have two rivaling ways to lint / format the code. Theres ESLint (`npm run lint`) and theres the whole gulp+jshint+esformatter setup  which @chrisprice introduced in #7). 

I guess my preferred way would be to retire the gulp setup and rely on eslint. It can do some formatting on itselt (`npm run lint -- --fix`) and reduces configuration to one package instead of a whole gulp setup. 

Could you have a look at the failed lints and see if you can make them pass? If you find something totally feels "wrong" we could discuss here if we keep it that way. 

Additionally theres a test failing[1]. I haven't looked what happens there exactly but I guess its nothing big. 

```
1) embl
       Pipeline
         should output input buffer:
     TypeError: buffers[ii].copy is not a function
      at Buffers.slice (node_modules/buffers/index.js:161:21)
      at Buffers.copy (node_modules/buffers/index.js:42:17)
      at Buffers.slice (node_modules/buffers/index.js:161:21)
      at Buffers.copy (node_modules/buffers/index.js:42:17)
      at Buffers.slice (node_modules/buffers/index.js:161:21)
      at Buffers.toBuffer (node_modules/buffers/index.js:264:17)
      at EbmlEncoder.endTag (lib/ebml/encoder.js:128:39)
      at EbmlEncoder._transform (lib/ebml/encoder.js:29:14)
      at EbmlEncoder.Transform._read (_stream_transform.js:186:10)
      at EbmlEncoder.Transform._write (_stream_transform.js:174:12)
      at doWrite (_stream_writable.js:397:12)
      at writeOrBuffer (_stream_writable.js:383:5)
      at EbmlEncoder.Writable.write (_stream_writable.js:290:11)
      at EbmlDecoder.ondata (_stream_readable.js:639:20)
      at addChunk (_stream_readable.js:263:12)
      at readableAddChunk (_stream_readable.js:250:11)
      at EbmlDecoder.Readable.push (_stream_readable.js:208:10)
      at EbmlDecoder.Transform.push (_stream_transform.js:147:32)
      at EbmlDecoder.readContent (lib/ebml/decoder.js:166:14)
      at EbmlDecoder._transform (lib/ebml/decoder.js:41:52)
      at EbmlDecoder.Transform._read (_stream_transform.js:186:10)
      at EbmlDecoder.Transform._write (_stream_transform.js:174:12)
      at doWrite (_stream_writable.js:397:12)
      at writeOrBuffer (_stream_writable.js:383:5)
      at EbmlDecoder.Writable.write (_stream_writable.js:290:11)
      at Context.<anonymous> (test/pipeline.js:15:21)
```

OllieJones commented:
Hello, thanks for getting back to me.

On Sun, Apr 22, 2018 at 6:32 AM, Mark Schmale <notifications@github.com>
wrote:

> Hi @OllieJones <https://github.com/OllieJones>, thanks for the massiv
> addition to the project.
> Could you please have a look at the tests & lints, since the builds are
> currently failing.
>

Yes, I will do this and make corrections. I will also try to add tests
against the webm / Matroska suite of test files.

I have a question for you: I'm actually using your code browser-side as
well as server side. browserify generates code that handles class
inheritance (for the Stream stuff) wrongly, so I refactored "decoder.js" to
make it stand alone   (All the changes in my commit work in
ie11/edge/safari/chrome/firefox as well as node, after that refactoring).
What should I do with that work?

Thanks.
Oliver Jones

themasch commented:
So if I understand you correct theres more refactoring you did to decodes.js then whats included in this PR? 

If you wan't to contribute this code back to the project you could create a second PR (after this one is merged) which adds the changes based on the new version of master. 
That way we have to deal with smaller change sets and can seperate the "new features" change from the "make it work without browserify" change. 

Sounds good?

OllieJones commented:
Hi,

On Mon, Apr 23, 2018 at 10:18 AM, Mark Schmale <notifications@github.com>
wrote:

> So if I understand you correct theres more refactoring you did to
> decodes.js then whats included in this PR?
>
Yes, that is right.

But it's API-incompatible with the existing decoder.js code. I will do a
pull request with my best effort at it, and you can give me advice about
how to make it fit in the project.

Thanks.  (And thanks for saving my **s with your fine code. It would have
taken me a long time to replicate it, and I probably never would have
debugged it successfully.)
-Ollie Jones

themasch commented:
@OllieJones to be honest, most of this code comes from other contributers just like you. I think theres not much left from my original code base.

OllieJones commented:
Hello, I've corrected the failing test and the lint problems. This should be ready to go.

themasch commented on .eslintrc.yml:
```diff
@@ -98,6 +98,7 @@ rules:
   no-caller: error
   no-catch-shadow: error
   no-confusing-arrow: error
+  no-console: 'off'
```
I don't think you need to disable this role. eslint seems to pass without this line since it lints `lib/` and `index.js` only and only the examples contain console.logs.

themasch commented on lib/ebml/tools.js:
```diff
@@ -50,6 +63,161 @@ var tools = {
+    },
+    readUtf8: function(buff) {
+        if (typeof window === 'undefined') {
+            return new Buffer.from(buff).toString("utf8");
```
As far as I know (I might be wrong here) the `new` is not needed (or even wrong?) here since `Buffer.from` creates a buffer itself.

themasch commented:
Wow, thanks! Thats some great work there. I annotated two minor things I recognized. Could you please have a look at them?

OllieJones commented:
Hello again, Mark Schmale.  I fixed the error using Buffer, sorted out the eslint problems, and added a test module.

This looks like it passes all the CI tests. 

Thanks for your patience!
